### ISSUE TITLE
Docker aufs issues when running lots of `pull` and `run` commands in parallel

### ISSUE THREAD

**Issue by zviadm** [docker/docker#16376]
I am running docker in a test environment, where there are multiple worker processes continuously pulling and running different containers.

After some point, some of the machines start throwing errors like this one:
```
Error response from daemon: error creating aufs mount to /mnt/docker/aufs/mnt/f2fca0b7ba13476a3f6b89ac986c9eae7e3876640dbe641439218a89c5875540-init: invalid argument
```

Or, this one:
```
EXEC[docker]: [pull XXXX:XXXX/isotester:main]
main: Pulling from isotester
3c2870d005ed: Pulling fs layer
3c2870d005ed: Pulling fs layer
9d276fcfbea2: Already exists
238ca624d2ea: Already exists
8615058f3e1f: Already exists
9610cfc68e8d: Already exists
7d152d6a4550: Already exists
eb86f0592ac6: Already exists
0f36a61b7011: Already exists
859b7f97b67f: Already exists
e960c44cc53d: Already exists
19b3b8ea9fdb: Already exists
be95702c4500: Already exists
c22e383b9bf7: Already exists
fb68523634d7: Already exists
ff621d625f79: Already exists
323ccbe757ac: Already exists
c515664bbada: Already exists
7d08fcb10882: Already exists
9b8b262231c6: Already exists
cd91592a40ca: Already exists
07aee4985daa: Already exists
7118b7235dc9: Already exists
102840b21d2c: Already exists
3c2870d005ed: Layer already being pulled by another client. Waiting.
3c2870d005ed: Verifying Checksum
3c2870d005ed: Download complete
3c2870d005ed: Download complete
Pulling repository XXXX/isotester
Error: image isotester:main not found
```

Restarting docker daemon makes the issues go away. Until it gets into this state again after few hours or days of running.

```
uname -a
Linux XXXX 3.16.0-33-generic #44~14.04.1-Ubuntu SMP Fri Mar 13 10:33:29 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

```
docker info
Containers: 2
Images: 78
Storage Driver: aufs
 Root Dir: /mnt/docker/aufs
 Backing Filesystem: xfs
 Dirs: 87
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-33-generic
Operating System: Ubuntu precise (12.04.5 LTS)
CPUs: 32
Total Memory: 58.97 GiB
Name: XXXX
ID: 26V6:LVAB:DJ4R:GERJ:KS2T:IT3O:VLBE:KZUI:RF5I:57FF:BUCO:P3CU
WARNING: No swap limit support
```

Running on EC2 boxes.

Following issues also seem potentially related:
https://github.com/kubernetes/kubernetes/issues/10959
https://github.com/docker/docker/issues/13742

**Comment by zviadm** [comment]
I included and docker info, here is the version too:
```
sudo docker version
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:16:50 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:16:50 UTC 2015
 OS/Arch:      linux/amd64
```


Steps to repro are quite hard to describe. Mostly I am suspecting this is due to lots of parallel `docker pull` and `docker run` commands happening on the same machine. Where `docker pull` is actually pulling new versions of an image that is changing in the registry.

**Comment by zviadm** [comment]
As I suspected synchronizing `docker pull` calls application side has fixed the issue. `docker run`-s seem to be safe to run in parallel but there has to be some sort of race/bugs in running bunch of `docker pull` requests in parallel.

I would suspect, most likely repro of this in a test case would be to setup different images like:
myrepo:test1
myrepo:test2
myrepo:test3

That change (i.e. get docker pushed continuously) and have some sort of dependencies on each other , for example: `myrepo:test2` is an image that is built from `myrepo:test1` and `myrepo:test3` is built from `myrepo:test2`.

And have a stress testing setup that runs a lot of docker pulls on those three images in parallel while they are continuously being pushed.

I would suspect that you hsould be able to reproduce most of the races that I was running into that way.

**Comment by LK4D4** [comment]
@zviadm There are #15283 and #9718 which looks similar. You might want to try master also, where some races in pull code were fixed.

**Comment by zviadm** [comment]
ah thanks @LK4D4 it does look very similar. They can definitely be merged into the same issue. Just for completeness here is  also list of issues that I have run into, which no longer seem to happen once I synchronized all `docker pull` calls:


docker pull error-ing out with `Error: image XXX not found` this was probably most common one
docker pull getting stuck forever, even though no other docker pulls would be running at that point

docker pull requests succeeding but now docker run commands would through different type of AUFS related errors, like:
  * Error response from daemon: error creating aufs mount to /mnt/docker/aufs/mnt/XXX-init: invalid argument
  * Error response from daemon: open /mnt/docker/aufs/layers/XXX: no such file or directory

Some of them actually stayed permanent even after restarting docker daemon, until I would push new version of the image to the registry and pulled the new stuff down to change whatever corrupted local state there was inside the /aufs folder.

**Comment by thaJeztah** [comment]
Closing as duplicate of #15283 (and #9718), please continue the discussion on those issues, thanks!
